js_modules/dagit/packages/core/src/runs/ReexecutionDialog.tsx

```diff
@@ -0,0 +1,343 @@
+      return 'Run config invalid';
+    case 'RunConflict':
+      return 'Run conflict';
+    case 'UnauthorizedError':
+      return 'Re-execution not authorized';
+    default:
+      return 'Unknown error';
+  }
+};
+
+export type ReexecutionState = {completed: number; errors: {[id: string]: Error}};
+
+type ReexecutionDialogState = {
+  frozenRuns: SelectedRuns;
+  step: 'initial' | 'reexecuting' | 'completed';
+  reexecution: ReexecutionState;
+};
+
+type SelectedRuns = {[id: string]: RunTableRunFragment};
+
+const initializeState = (selectedRuns: SelectedRuns): ReexecutionDialogState => {
+  return {
+    frozenRuns: selectedRuns,
+    step: 'initial',
+    reexecution: {completed: 0, errors: {}},
+  };
+};
+
+type ReexecutionDialogAction =
+  | {type: 'reset'; frozenRuns: SelectedRuns}
+  | {type: 'start'}
+  | {type: 'reexecution-success'}
+  | {type: 'reexecution-error'; id: string; error: Error}
+  | {type: 'complete'};
+
+const reexecutionDialogReducer = (
+  prevState: ReexecutionDialogState,
+  action: ReexecutionDialogAction,
+): ReexecutionDialogState => {
+  switch (action.type) {
+    case 'reset':
+      return initializeState(action.frozenRuns);
+    case 'start':
+      return {...prevState, step: 'reexecuting'};
+    case 'reexecution-success': {
+      const {reexecution} = prevState;
+      return {
+        ...prevState,
+        step: 'reexecuting',
+        reexecution: {...reexecution, completed: reexecution.completed + 1},
+      };
+    }
+    case 'reexecution-error': {
+      const {reexecution} = prevState;
+      return {
+        ...prevState,
+        step: 'reexecuting',
+        reexecution: {
+          ...reexecution,
+          completed: reexecution.completed + 1,
+          errors: {...reexecution.errors, [action.id]: action.error},
+        },
+      };
+    }
+    case 'complete':
+      return {...prevState, step: 'completed'};
+  }
+};
+
+export const ReexecutionDialog = (props: Props) => {
+  const {isOpen, onClose, onComplete, selectedRuns} = props;
+
+  // Freeze the selected IDs, since the list may change as runs continue processing and
+  // re-executing. We want to preserve the list we're given.
+  const frozenRuns = React.useRef<SelectedRuns>(selectedRuns);
+
+  const [state, dispatch] = React.useReducer(
+    reexecutionDialogReducer,
+    frozenRuns.current,
+    initializeState,
+  );
+
+  const repositoryMatches = useRepositoriesForRuns(state.frozenRuns);
+
+  const count = Object.keys(state.frozenRuns).length;
+
+  // If the dialog is newly open, update state to match the frozen list.
+  React.useEffect(() => {
+    if (isOpen) {
+      dispatch({type: 'reset', frozenRuns: frozenRuns.current});
+    }
+  }, [isOpen]);
+
+  // If the dialog is not open, update the ref so that the frozen list will be entered
+  // into state the next time the dialog opens.
+  React.useEffect(() => {
+    if (!isOpen) {
+      frozenRuns.current = selectedRuns;
+    }
+  }, [isOpen, selectedRuns]);
```

Comment:
(Looks good to me but this ref + effect might be another argument for the "detatched dialogs" pitch ;-) )